hash_writer.go

```diff
@@ -33,157 +33,249 @@ type HashWriter interface {
 }
 
 type hashWriter struct {
-	ionWriter      ion.Writer
-	hasherProvider IonHasherProvider
+	ionWriter ion.Writer
+	hasher    hasher
+
+	currentFieldName string
+	currentType      ion.Type
+	currentValue     interface{}
+	currentIsNull    bool
+	annotations      []string
 }
 
 func NewHashWriter(ionWriter ion.Writer, hasherProvider IonHasherProvider) HashWriter {
-	hashWriter := &hashWriter{ionWriter, hasherProvider}
+	hasher := newHasher(hasherProvider)
+	hashWriter := &hashWriter{ionWriter: ionWriter, hasher: *hasher}
 
 	return hashWriter
 }
 
 func (hashWriter *hashWriter) FieldName(val string) error {
-	hashWriter.ionWriter.FieldName(val)
-	panic("implement me")
+	hashWriter.currentFieldName = val
+	return hashWriter.ionWriter.FieldName(val)
 }
 
 func (hashWriter *hashWriter) Annotation(val string) error {
```

Comment:
Since Annotations can be more than one, probably we can change this ```func (hashWriter *hashWriter) Annotations(val ...string) error {```
It would behave the same whether it's called with one or more than one annotations.

Reply:
Oh nevermind! I saw the next function!

Reply:
Why have both? Why not just use the general Annotations(val ...string) function to handle 1 or more annotations?

Reply:
We need both functions because of the writer interface